src/Std/Time/Zoned/ZonedDateTime.lean

```diff
@@ -0,0 +1,515 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sofia Rodrigues
+-/
+prelude
+import Std.Time.Zoned.DateTime
+import Std.Time.Zoned.ZoneRules
+
+namespace Std
+namespace Time
+
+set_option linter.all true
+
+/--
+Represents a date and time with timezone information.
+-/
+structure ZonedDateTime where
+  private mk::
+
+  /--
+  The plain datetime component, evaluated lazily.
+  -/
+  date : Thunk PlainDateTime
+
+  /--
+  The corresponding timestamp for the datetime.
+  -/
+  timestamp : Timestamp
+
+  /--
+  The timezone rules applied to this datetime.
+  -/
+  rules : TimeZone.ZoneRules
+
+  /--
+  The timezone associated with this datetime.
+  -/
+  timezone : TimeZone
+
+instance : Inhabited ZonedDateTime where
+  default := ⟨Thunk.mk Inhabited.default, Inhabited.default, Inhabited.default, Inhabited.default⟩
+
+namespace ZonedDateTime
+open DateTime
+
+/--
+Creates a new `ZonedDateTime` out of a `Timestamp` and a `ZoneRules`.
+-/
+@[inline]
+def ofTimestamp (tm : Timestamp) (rules : TimeZone.ZoneRules) : ZonedDateTime :=
+  let tz := rules.timezoneAt tm
+  ZonedDateTime.mk (Thunk.mk fun _ => tm.toPlainDateTimeAssumingUTC.addSeconds tz.toSeconds) tm rules tz
+
+/--
+Creates a new `ZonedDateTime` out of a `PlainDateTime` and a `ZoneRules`.
+-/
+@[inline]
+def ofPlainDateTime (pdt : PlainDateTime) (zr : TimeZone.ZoneRules) : ZonedDateTime :=
```

Comment:
```lean
import Std.Time

open Std.Time

def main : IO Unit := do
  let zr ← Database.defaultGetZoneRules "America/Sao_Paulo"
  let pdt := datetime("2019-02-17T00:00:01")
  let zdt := ZonedDateTime.ofPlainDateTime pdt zr
  println! "{zdt}"
```
This prints `2019-02-17T00:00:01.000000000-02:00`, but the correct answer is `2019-02-17T00:00:01.000000000-03:00` (Sao Paulo left DST, for the final time, one second earlier).

I knew that this function must be wrong because it called `toTimestampAssumingUTC` on a `PlainDateTime` that is not in UTC. This is what I was alluding to in [this comment](https://github.com/leanprover/lean4/pull/4904#discussion_r1803217660). Treating a `Timestamp` as a point on the timeline and never having a `Timestamp` that represents something without meaning (like `pdt.toTimestampAssumingUTC` if `pdt` is not in UTC) is a good way to avoid bugs like this.

Reply:
Fixed!